### PR TITLE
Rewrite balanced host attribute test

### PR DESCRIPTION
**Pull request by pschorf**

### CODE REVIEW

File: integration/tests/cook/test_basic.py [hunk 1]
```diff
@@ -997,23 +997,56 @@ def check_unique_constraint(unscheduled_jobs):
...
+                reason = reasons[0]
+                balanced_reasons = [r for r in reason['data']['reasons']
+                                    if r['reason'] == 'balanced-host-placement-group-constraint']
+                self.assertEqual(1, len(balanced_reasons), balanced_reasons)
+            # Otherwise, all jobs should be scheduled. Just check that the distribution is balanced.
+            else:
+                jobs = util.wait_for_jobs(self.cook_url, uuids, 'running')
+                hosts = [j['instances'][0]['hostname'] for j in jobs]
+                host_count = Counter(hosts)
+                self.assertTrue(len(host_count) >= minimum_hosts)
```
Comment (shamsimam) [inline integration/tests/cook/test_basic.py]: Nitpick: please use `assertGreaterEqual` https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertGreater

File: integration/tests/cook/test_basic.py [hunk 2]
```diff
@@ -997,23 +997,56 @@ def check_unique_constraint(unscheduled_jobs):
...
+                self.assertTrue(max(host_count.values()) <= max_jobs_per_host, host_count)
```
Comment (shamsimam) [inline integration/tests/cook/test_basic.py]: Nitpick: please use `assertLessEqual` https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertLessEqual

File: integration/tests/cook/test_basic.py [hunk 3]
```diff
@@ -997,23 +997,56 @@ def check_unique_constraint(unscheduled_jobs):
     def test_balanced_host_constraint(self):
```
Comment (shamsimam) [inline integration/tests/cook/test_basic.py]: Change requested: Can we split this test into two tests, one that runs for a small cluster (`minimum_hosts == num_hosts + 1`) and one that runs for a larger cluster.

File: integration/tests/cook/test_basic.py [hunk 4]
```diff
@@ -997,23 +997,56 @@ def check_unique_constraint(unscheduled_jobs):
...
+                    return util.query_jobs(self.cook_url, job=uuids).json()
+                def num_running_predicate(response):
+                    num_running = len([j for j in response if j['status'] == 'running'])
+                    num_waiting = len([j for j in response if j['status'] == 'waiting'])
+                    return num_running == num_hosts and num_waiting == 1
+                jobs = util.wait_until(query, num_running_predicate)
+                waiting_jobs = [j for j in jobs if j['status'] == 'waiting']
+                self.assertEqual(1, len(waiting_jobs), waiting_jobs)
+                waiting_job = waiting_jobs[0]
+                def query():
```
Comment (shamsimam) [inline integration/tests/cook/test_basic.py]: Nitpick: can we use different names for the two `query` functions?

File: integration/tests/cook/test_basic.py [hunk 5]
```diff
@@ -994,16 +994,60 @@ def check_unique_constraint(unscheduled_jobs):
         finally:
             util.kill_jobs(self.cook_url, [uuids])
 
+    def test_balanced_host_constraint_cannot_place(self):
+        state = util.get_mesos_state(self.mesos_url)
+        num_hosts = len(state['slaves'])
+        if num_hosts > 10:
+            # Skip this test on large clusters
```
Comment (shamsimam) [inline integration/tests/cook/test_basic.py]: Nitpick, can we add a log specifying we are skipping this test and also mention the cluster size

File: integration/tests/cook/test_basic.py [hunk 6]
```diff
@@ -994,16 +994,60 @@ def check_unique_constraint(unscheduled_jobs):
...
+            return
+        minimum_hosts = num_hosts + 1
+        group = {'uuid': str(uuid.uuid4()),
+                 'host-placement': {'type': 'balanced',
+                                    'parameters': {'attribute': 'HOSTNAME',
+                                                   'minimum': minimum_hosts}}}
+        job_spec = {'group': group['uuid'],
+                    'command': 'sleep 600',
+                    'mem': 100,
+                    'cpus': 0.1}
```
Comment (shamsimam) [inline integration/tests/cook/test_basic.py]: Nitpick: lexicographic ordering

File: integration/tests/cook/test_basic.py [hunk 7]
```diff
@@ -994,16 +994,60 @@ def check_unique_constraint(unscheduled_jobs):
...
             util.kill_jobs(self.cook_url, [uuids])
 
+    def test_balanced_host_constraint_cannot_place(self):
+        state = util.get_mesos_state(self.mesos_url)
+        num_hosts = len(state['slaves'])
+        if num_hosts > 10:
+            # Skip this test on large clusters
+            return
+        minimum_hosts = num_hosts + 1
+        group = {'uuid': str(uuid.uuid4()),
```
Comment (shamsimam) [inline integration/tests/cook/test_basic.py]: Nitpick: recursive lexicographic ordering

File: integration/tests/cook/test_basic.py [hunk 8]
```diff
@@ -994,16 +994,60 @@ def check_unique_constraint(unscheduled_jobs):
         finally:
             util.kill_jobs(self.cook_url, [uuids])
 
+    def test_balanced_host_constraint_cannot_place(self):
+        state = util.get_mesos_state(self.mesos_url)
+        num_hosts = len(state['slaves'])
+        if num_hosts > 10:
+            # Skip this test on large clusters
+            return
+        minimum_hosts = num_hosts + 1
```
Comment (shamsimam) [inline integration/tests/cook/test_basic.py]: Nitpick: Do we need a separate `minimum_hosts` variable?

File: integration/tests/cook/test_basic.py (hunk 5 above)
Comment (pschorf) [reply integration/tests/cook/test_basic.py]: Done

File: integration/tests/cook/test_basic.py (hunk 8 above)
Comment (pschorf) [reply integration/tests/cook/test_basic.py]: We don't need one, but I think it adds to clarity

File: integration/tests/cook/test_basic.py (hunk 7 above)
Comment (pschorf) [reply integration/tests/cook/test_basic.py]: I think this is already sorted in a sensible manner.

File: integration/tests/cook/test_basic.py (hunk 6 above)
Comment (pschorf) [reply integration/tests/cook/test_basic.py]: done

File: integration/tests/cook/test_basic.py (hunk 8 above)
Comment (shamsimam) [reply integration/tests/cook/test_basic.py]: okay

**Comment by pschorf** [comment]
Build is green

**Pull request merged**